lib/internal/test_runner/test.js

```diff
@@ -76,15 +75,40 @@ const { testNamePatterns, testOnlyFlag } = parseCommandLine();
 let kResistStopPropagation;
 
 function stopTest(timeout, signal) {
+  const deferred = createDeferredPromise();
+
   if (timeout === kDefaultTimeout) {
-    return once(signal, 'abort');
+    signal.addEventListener('abort', deferred.resolve, { __proto__: null, once: true });
+    deferred.cleanup = () => signal.removeEventListener('abort', deferred.resolve);
+    return deferred;
+  }
+
+  const timer = setTimeout(() => {
+    deferred.resolve();
+    signal.removeEventListener('abort', throwOnAbort);
+  }, timeout);
+
+  timer.unref();
+  function throwOnAbort() {
+    deferred.reject(new AbortError(undefined, { cause: signal.reason }));
+    clearTimeout(timer);
   }
-  return PromisePrototypeThen(setTimeout(timeout, null, { ref: false, signal }), () => {
+
+  signal.addEventListener('abort', throwOnAbort, { __proto__: null, once: true });
+
+  deferred.cleanup = () => {
+    clearTimeout(timer);
+    signal.removeEventListener('abort', throwOnAbort);
+  };
+
+  deferred.promise = PromisePrototypeThen(deferred.promise, () => {
     throw new ERR_TEST_FAILURE(
       `test timed out after ${timeout}ms`,
       kTestTimeoutFailure,
     );
   });
```

Comment:
I wanted to put this in the `setTimeout` but for some reason, the `test-runner-output.mjs` test failed for the `describe_it` file with some missing stack trace